## Problem Overview

We perform sentiment analysis on the TweetEval dataset using the same model
(DistilBERT) under three settings:

1. Baseline (zero-shot / no training)
2. In-Context Learning (few-shot prompting)
3. Fine-tuning (supervised learning)

### Dataset Overview

- **Source**: TweetEval Sentiment benchmark
- **Task**: 3-class sentiment classification (Negative, Neutral, Positive)
- **Challenge**: Class imbalance in real-world Twitter data


In [3]:
!pip install -q transformers datasets evaluate accelerate scikit-learn torch

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 4.5 MB/s eta 0:00:00


### Dependencies Installed

- `transformers`: Hugging Face library for pretrained models
- `datasets`: Dataset loading and processing
- `evaluate`: Metrics computation (F1, accuracy)
- `accelerate`: Training optimization
- `scikit-learn`: Class weight computation, evaluation metrics
- `torch`: Deep learning framework (PyTorch)

All packages installed successfully! 

In [12]:
import torch
import numpy as np
import random

from datasets import load_dataset
from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    Trainer,
    TrainingArguments,
    pipeline
)

from sklearn.metrics import classification_report
from sklearn.utils.class_weight import compute_class_weight
import evaluate

### Environment Setup

**Libraries Imported**:
- **Model & Training**: `transformers` (AutoTokenizer, AutoModel, Trainer)
- **Data**: `datasets` (load_dataset)
- **Evaluation**: `sklearn.metrics`, `evaluate`
- **Utilities**: `torch`, `numpy`, `random`

**GPU Check**:
```python
import torch
print(f"CUDA Available: {torch.cuda.is_available()}")
print(f"Device: {torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'CPU'}")
```

## Dataset Loading & Analysis

TweetEval (Sentiment) has 3 labels:
- 0 → Negative
- 1 → Neutral
- 2 → Positive

We first analyze class distribution to confirm imbalance.

In [13]:
dataset = load_dataset("tweet_eval", "sentiment")

## Tokenization

We tokenize all splits using DistilBERT tokenizer.

In [14]:
tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")

def tokenize(batch):
    return tokenizer(
        batch["text"],
        padding="max_length",
        truncation=True,
        max_length=128
    )

dataset = dataset.map(tokenize, batched=True)

dataset = dataset.remove_columns(
    [c for c in dataset["train"].column_names if c not in ["input_ids", "attention_mask", "label"]]
)

dataset.set_format("torch")

Map:   0%|          | 0/45615 [00:00<?, ? examples/s]

Map:   0%|          | 0/12284 [00:00<?, ? examples/s]

Map:   0%|          | 0/2000 [00:00<?, ? examples/s]

### Tokenization Complete

**What Happened**:
1. Converted text to token IDs using DistilBERT tokenizer
2. Applied padding to max_length=128 (sufficient for tweets)
3. Applied truncation for longer texts
4. Generated attention masks

## Fixed Dataset Subsets

To ensure fast and fair comparison, we create fixed-size subsets
using shuffle + select with a fixed random seed.

In [15]:
train_data = dataset["train"].shuffle(seed=42).select(range(5000))
val_data   = dataset["validation"].shuffle(seed=42).select(range(1000))
test_data  = dataset["test"].shuffle(seed=42).select(range(1000))

print(f"Train: {len(train_data)}")
print(f"Val:   {len(val_data)}")
print(f"Test:  {len(test_data)}")

Train: 5000
Val:   1000
Test:  1000


### Fixed Dataset Subsets Created

**Subset Sizes**:
- Training: 5,000 samples (from 45,615)
- Validation: 1,000 samples (from 12,284)
- Test: 1,000 samples (from 2,000)

**Why Use Subsets?**:
1. **Computational Efficiency**: Faster training and evaluation
2. **Fair Comparison**: All approaches use identical data
3. **Reproducibility**: Fixed seed (42) ensures consistent splits
4. **Representative**: Random sampling preserves distribution


#  Approach 1: Baseline (Zero-Shot)

### Concept
Use pretrained DistilBERT **without any training** on TweetEval data. This tests how well the model generalizes from its original pretraining.

### Expected Behavior
- Model has pretrained language understanding
- **But**: Classification head is randomly initialized
- **Result**: Likely to be biased toward majority class or random predictions

### Why Test This?
- Establishes lower bound performance
- Shows importance of task-specific training
- Tests transfer learning capabilities

In [16]:
baseline_model = AutoModelForSequenceClassification.from_pretrained(
    "distilbert-base-uncased",
    num_labels=3
)

baseline_model.eval()

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


DistilBertForSequenceClassification(
  (distilbert): DistilBertModel(
    (embeddings): Embeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (transformer): Transformer(
      (layer): ModuleList(
        (0-5): 6 x TransformerBlock(
          (attention): DistilBertSdpaAttention(
            (dropout): Dropout(p=0.1, inplace=False)
            (q_lin): Linear(in_features=768, out_features=768, bias=True)
            (k_lin): Linear(in_features=768, out_features=768, bias=True)
            (v_lin): Linear(in_features=768, out_features=768, bias=True)
            (out_lin): Linear(in_features=768, out_features=768, bias=True)
          )
          (sa_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (ffn): FFN(
            (dropout): Dropout(p=0.1, inplace=False)


### Baseline Model Loaded

**Model**: `distilbert-base-uncased`
- **Parameters**: ~66M (40% smaller than BERT)
- **Layers**: 6 transformer layers
- **Hidden size**: 768
- **Classification head**: Randomly initialized for 3 classes

**Warning About Random Weights**:
The warning "some weights were not initialized" is **expected** - the classification head (pre_classifier and classifier layers) are randomly initialized since the pretrained model wasn't trained for our specific task.

In [17]:
from torch.utils.data import DataLoader

def evaluate_model(model, dataset, batch_size=32):
    loader = DataLoader(dataset, batch_size=batch_size, shuffle=False)

    preds, labels = [], []

    with torch.no_grad():
        for batch in loader:
            outputs = model(
                input_ids=batch["input_ids"],
                attention_mask=batch["attention_mask"]
            )
            batch_preds = torch.argmax(outputs.logits, dim=1)

            preds.extend(batch_preds.cpu().numpy())
            labels.extend(batch["label"].cpu().numpy())

    return labels, preds

###  Evaluation Function Defined

**How It Works**:
```python
def evaluate_model(model, dataset, batch_size=32):
    # 1. Creates DataLoader for batching
    # 2. Disables gradients (no training)
    # 3. Iterates through batches
    # 4. Collects predictions and true labels
    # 5. Returns for metric calculation
```

**Why Batch Processing?**:
- More efficient than one-by-one
- Better GPU utilization
- Faster inference time

**Key Detail**: Uses `torch.no_grad()` to prevent memory buildup during inference.

In [18]:
y_true_base, y_pred_base = evaluate_model(baseline_model, test_data)

print(classification_report(y_true_base, y_pred_base, digits=4))

              precision    recall  f1-score   support

           0     0.3293    0.0816    0.1308       331
           1     0.4749    0.9179    0.6260       475
           2     0.0000    0.0000    0.0000       194

    accuracy                         0.4630      1000
   macro avg     0.2681    0.3332    0.2522      1000
weighted avg     0.3346    0.4630    0.3406      1000



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


### Baseline Results Analysis

**Performance**:
Accuracy:  46.3%
Macro-F1:  25.2%

**Per-Class Breakdown**:
| Class | Precision | Recall | F1-Score | Issue |
|-------|-----------|--------|----------|-------|
| Negative | 32.9% | 8.2% | 13.1% | Barely detected |
| Neutral | 47.5% | 91.8% | 62.6% | **Heavily biased** |
| Positive | 0.0% | 0.0% | 0.0% | **Completely missed** |

**Critical Observations**:

1. **Majority Class Bias**: Model predicts Neutral 91.8% of the time
2. **Minority Class Failure**: Cannot detect Positive sentiment at all
3. **Random Classification Head**: Untrained layer defaults to safe predictions
4. **Imbalance Effect**: Without training, model gravitates toward most common class

**Why Accuracy is Misleading**:
- 46.3% accuracy seems "okay"
- But Macro-F1 of 25.2% reveals true poor performance
- **Lesson**: Always use class-balanced metrics for imbalanced data!

**Conclusion**: Zero-shot baseline fails due to random classification head and class imbalance.

---

# Approach 2: In-Context Learning (Few-Shot)

### Concept
Provide labeled examples in the prompt to "teach" the model through context, without parameter updates.

### How It Works

Tweet: I hate this movie. Sentiment: Negative.
Tweet: This is okay. Sentiment: Neutral.
Tweet: I love this phone. Sentiment: Positive.
Tweet: [TEST_TWEET]. Sentiment: [PREDICT]

### Hypothesis
ICL works well for large language models (GPT-3/4) that can follow instructions. But DistilBERT is an **encoder-only** model designed for classification, not instruction-following.

**Expected Outcome**: Limited or no improvement, possibly worse than baseline.

### Why Test This?
- Popular technique in modern NLP (GPT era)
- Tests architectural compatibility
- Shows when ICL is/isn't appropriate

In [19]:
FEW_SHOT_EXAMPLES = [
    ("I hate this movie", "Negative"),
    ("This is okay", "Neutral"),
    ("I love this phone", "Positive")
]

def create_prompt(text):
    prompt = ""
    for ex, label in FEW_SHOT_EXAMPLES:
        prompt += f"Tweet: {ex}. Sentiment: {label}.\n"
    prompt += f"Tweet: {text}. Sentiment:"
    return prompt

### Few-Shot Prompt Template

**Prompt Structure**:
1. Three labeled examples (one per class)
2. Clear formatting: "Tweet: X. Sentiment: Y."
3. Test tweet at the end
4. Model predicts sentiment label

**Design Choices**:
- Balanced examples (one per class)
- Simple, clear language
- Explicit sentiment labels
- Consistent formatting

**Limitation**: DistilBERT processes this as one long text, not as structured instructions.

In [20]:
icl_pipe = pipeline(
    "text-classification",
    model=baseline_model,
    tokenizer=tokenizer
)

label_map = {"NEGATIVE": 0, "NEUTRAL": 1, "POSITIVE": 2}

Device set to use cuda:0


In [21]:
y_true_icl, y_pred_icl = [], []

for sample in test_data:
    prompt = create_prompt(sample["input_ids"])  # token ids not used here
    text = tokenizer.decode(sample["input_ids"], skip_special_tokens=True)

    pred = icl_pipe(create_prompt(text))[0]["label"]

    y_true_icl.append(sample["label"].item())
    y_pred_icl.append(label_map.get(pred, 1))

print(classification_report(y_true_icl, y_pred_icl, digits=4))

You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset


              precision    recall  f1-score   support

           0     0.0000    0.0000    0.0000       331
           1     0.4750    1.0000    0.6441       475
           2     0.0000    0.0000    0.0000       194

    accuracy                         0.4750      1000
   macro avg     0.1583    0.3333    0.2147      1000
weighted avg     0.2256    0.4750    0.3059      1000



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


### ICL Results Analysis

**Performance**:

Accuracy:  47.5%
Macro-F1:  21.5%

**Per-Class Breakdown**:
| Class | Precision | Recall | F1-Score | Issue |
|-------|-----------|--------|----------|-------|
| Negative | 0.0% | 0.0% | 0.0% | Not detected |
| Neutral | 47.5% | 100% | 64.4% | **Only prediction** |
| Positive | 0.0% | 0.0% | 0.0% | Not detected |

**Critical Finding: ICL FAILED**

**Why ICL Performed WORSE Than Baseline**:

1. **Architecture Mismatch**: 
   - DistilBERT is encoder-only (designed for classification)
   - ICL requires decoder models (GPT) that generate text
   - Encoder models don't "understand" instruction-following

2. **Prompt Confusion**:
   - Extra examples add noise, not guidance
   - Classification head trained on single tweets, not multi-example prompts
   - Token space wasted on examples instead of test tweet

3. **No Learning Mechanism**:
   - ICL relies on model adapting to examples
   - DistilBERT has no mechanism to update from context
   - Predictions remain random

4. **Even More Biased**:
   - Now predicts **only Neutral** (100% recall)
   - Completely ignores Negative and Positive

**Key Lesson**: 
**In-Context Learning is NOT suitable for encoder-only models like DistilBERT.**
 
 Use ICL with: GPT-3/4, Claude, T5 (decoder or encoder-decoder models)
 
 Use Fine-tuning with: BERT, DistilBERT, RoBERTa (encoder models)

**This negative result is valuable** - it shows when NOT to use ICL!

---

# Approach 3: Fine-Tuning (Expected Best)

### Concept
Train DistilBERT on labeled TweetEval data with **class-weighted loss** to handle imbalance.

### Why This Should Work
1. **Supervised Learning**: Model updates parameters based on labeled data
2. **Task-Specific Adaptation**: Classification head learns sentiment patterns
3. **Imbalance Handling**: Class weights force model to learn minority classes
4. **Architecture Match**: Encoder models designed for fine-tuning

### Key Innovation: Class-Weighted Loss

**Problem**: Standard loss treats all classes equally
```python
# Standard loss
loss = CrossEntropyLoss()(predictions, labels)
# Result: Model biased toward majority class
```

**Solution**: Weight loss by inverse class frequency
```python
# Weighted loss
class_weights = [2.15, 0.73, 0.85]  # Higher for minority classes
loss = CrossEntropyLoss(weight=class_weights)(predictions, labels)
# Result: Model penalized more for missing minority classes
```

### Expected Outcome
- Balanced performance across all classes
- Significant improvement over baseline
- Best macro-F1 score

In [24]:
# Convert labels to numpy array
labels = np.array(train_data["label"])

# Explicitly specify all possible classes
class_weights = compute_class_weight(
    class_weight="balanced",
    classes=np.array([0, 1, 2]),
    y=labels
)

class_weights = torch.tensor(class_weights, dtype=torch.float)

print("Class weights:", class_weights)

Class weights: tensor([2.1505, 0.7345, 0.8521])


### Class Weights Computed

**Computed Weights**:
```python
Class 0 (Negative): 2.1505
Class 1 (Neutral):  0.7345
Class 2 (Positive): 0.8521
```

**Interpretation**:
- **High weight (2.15)** → Negative is **underrepresented** (minority class)
- **Low weight (0.73)** → Neutral is **overrepresented** (majority class)
- **Medium weight (0.85)** → Positive is slightly overrepresented

**What This Does**:

For Negative samples: loss × 2.15 → Model penalized MORE for mistakes

For Neutral samples:  loss × 0.73 → Model penalized LESS for mistakes

For Positive samples: loss × 0.85 → Standard penalty

**Effect on Training**:
- Model must work harder to correctly classify Negative tweets
- Prevents collapse to majority class (Neutral)
- Encourages balanced learning across all classes

**Formula**:

weight_i = n_total / (n_classes × n_samples_i)

Where:
- `n_total` = 5,000 (total training samples)
- `n_classes` = 3
- `n_samples_i` = number of samples in class i

In [ ]:
class WeightedTrainer(Trainer):
    def compute_loss(
        self,
        model,
        inputs,
        return_outputs=False,
        num_items_in_batch=None  
    ):
        labels = inputs.get("labels")
        outputs = model(**inputs)
        logits = outputs.get("logits")

        loss_fn = torch.nn.CrossEntropyLoss(weight=class_weights.to(logits.device))
        loss = loss_fn(logits, labels)

        return (loss, outputs) if return_outputs else loss


### Custom Trainer with Weighted Loss

**Why Custom Trainer?**

The default Hugging Face `Trainer` uses standard CrossEntropyLoss. We override `compute_loss()` to inject class weights.

**Key Points**:
1. `num_items_in_batch` parameter required for newer transformers versions
2. Moves class weights to same device as model (GPU/CPU compatibility)
3. Returns both loss and outputs for gradient computation

**Alternative Approaches** (not used here):
- Oversampling minority classes
- SMOTE (synthetic data generation)
- Focal loss (focuses on hard examples)

We chose class weighting for its simplicity and effectiveness.

In [26]:
metric = evaluate.load("f1")

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    preds = np.argmax(logits, axis=1)
    return {
        "macro_f1": metric.compute(
            predictions=preds,
            references=labels,
            average="macro"
        )["f1"]
    }

### Evaluation Metric: Macro-F1

**Why Macro-F1?**

For imbalanced classification, we need metrics that treat all classes equally:

| Metric | Description | Problem with Imbalance |
|--------|-------------|----------------------|
| Accuracy | % correct predictions | Biased toward majority class |
| Weighted-F1 | F1 weighted by class size | Still favors majority |
| **Macro-F1** | Average F1 across classes | **Treats all classes equally** |

**Macro-F1 Calculation**:
Compute F1 for each class:
F1_neg = 2 × (precision_neg × recall_neg) / (precision_neg + recall_neg)
F1_neu = ...
F1_pos = ...

Average them:
Macro-F1 = (F1_neg + F1_neu + F1_pos) / 3

**Example**:
- Class A: F1 = 0.9 (1000 samples)
- Class B: F1 = 0.5 (100 samples)
- Class C: F1 = 0.3 (10 samples)

Weighted-F1 = (0.9×1000 + 0.5×100 + 0.3×10) / 1110 = 0.85

Macro-F1    = (0.9 + 0.5 + 0.3) / 3 = 0.57

Macro-F1 reveals that we're failing on minority classes!

**Our Goal**: Maximize Macro-F1 to ensure balanced performance.

In [28]:
training_args = TrainingArguments(
    output_dir="./results",
    eval_strategy="epoch",   # <-- FIX HERE
    save_strategy="no",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=3,
    weight_decay=0.01,
    logging_steps=100
)

### Training Configuration

**Hyperparameters**:
```python
learning_rate = 2e-5          # Standard for BERT-based models
batch_size = 16               # Balanced for GPU memory
epochs = 3                    # Typical for fine-tuning
weight_decay = 0.01           # L2 regularization
```

**Why These Values?**

1. **Learning Rate (2e-5)**:
   - Too high: Catastrophic forgetting of pretrained knowledge
   - Too low: Slow convergence
   - 2e-5 is standard for BERT fine-tuning (from original paper)

2. **Batch Size (16)**:
   - Larger = more stable gradients, needs more memory
   - Smaller = more updates, more noise
   - 16 is good balance for T4 GPU

3. **Epochs (3)**:
   - BERT-based models converge quickly
   - More epochs risk overfitting on small datasets
   - Validation will tell us if we need more

4. **Weight Decay (0.01)**:
   - Prevents overfitting
   - Regularizes large weight updates

**Evaluation Strategy**:
- `eval_strategy="epoch"`: Validate after each epoch
- Helps detect overfitting early
- Tracks learning progress

**Output**:
- Checkpoints: Disabled (`save_strategy="no"`) to save space
- Logs: Every 100 steps

### Starting Training...

**What to Watch For**:

1. **Training Loss**: Should decrease steadily
   - Epoch 1: High loss (model learning)
   - Epoch 2-3: Lower loss (convergence)

2. **Validation Macro-F1**: Should increase
   - Shows model improving on balanced metric
   - If it decreases, we're overfitting

3. **Training Time**:
   - Expected: ~2-3 minutes on T4 GPU
   - 5,000 samples × 3 epochs = 15,000 total samples
   - ~90 samples/second

**Training Progress Indicators**:
- `[XXX/939]`: Step counter (313 steps per epoch × 3)
- `Epoch X/3`: Current epoch
- `Loss`: Training loss at current step
- `Validation Loss`: Loss on validation set
- `Macro F1`: Our key metric!

In [31]:
finetune_model = AutoModelForSequenceClassification.from_pretrained(
    "distilbert-base-uncased",
    num_labels=3
)

trainer = WeightedTrainer(
    model=finetune_model,
    args=training_args,
    train_dataset=train_data,
    eval_dataset=val_data,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

trainer.train()

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/tmp/ipython-input-2772116460.py:6: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `WeightedTrainer.__init__`. Use `processing_class` instead.
  trainer = WeightedTrainer(


TrainOutput(global_step=939, training_loss=0.6341557832825552, metrics={'train_runtime': 167.3174, 'train_samples_per_second': 89.65, 'train_steps_per_second': 5.612, 'total_flos': 496761603840000.0, 'train_loss': 0.6341557832825552, 'epoch': 3.0})

### Training Complete!

**Training Summary**:
```python
Total Runtime: 167.32 seconds (2m 47s)
Training Speed: 89.65 samples/second
Steps per Second: 5.61
Total FLOPs: 4.97 × 10^14
```

**Learning Progress**:

| Epoch | Training Loss | Val Loss | Macro-F1 | Improvement |
|-------|---------------|----------|----------|-------------|
| 1 | 0.7365 | 0.7633 | 0.5962 | Baseline |
| 2 | 0.5782 | 0.7787 | 0.6422 | +7.7% |
| 3 | 0.4583 | 0.8301 | 0.6695 | +4.3% |

**Key Observations**:

1. **Training Loss Decreasing** 
   - 0.74 → 0.58 → 0.46
   - Model is learning!

2. **Validation Loss Increasing** 
   - 0.76 → 0.78 → 0.83
   - Slight overfitting, but Macro-F1 still improving
   - Not severe enough to stop

3. **Macro-F1 Steadily Improving** 
   - 0.60 → 0.64 → 0.67
   - Consistent progress across all epochs
   - No signs of plateauing

**Conclusion**: Training successful! Model learned effectively despite slight overfitting. Validation Macro-F1 improvement indicates genuine learning, not memorization.

**Next**: Evaluate on test set to see final performance!

In [32]:
preds = trainer.predict(test_data)

y_true_ft = preds.label_ids
y_pred_ft = np.argmax(preds.predictions, axis=1)

print(classification_report(y_true_ft, y_pred_ft, digits=4))

              precision    recall  f1-score   support

           0     0.6952    0.7372    0.7155       331
           1     0.7190    0.6358    0.6749       475
           2     0.5983    0.7062    0.6478       194

    accuracy                         0.6830      1000
   macro avg     0.6708    0.6930    0.6794      1000
weighted avg     0.6877    0.6830    0.6831      1000



### Fine-Tuning Results Analysis

**Final Performance**:

Accuracy:  68.3%

Macro-F1:  67.9%

**Per-Class Breakdown**:
| Class | Precision | Recall | F1-Score | vs Baseline |
|-------|-----------|--------|----------|-------------|
| Negative | 69.5% | 73.7% | 71.6% | +447%  |
| Neutral | 71.9% | 63.6% | 67.5% | +7.8% |
| Positive | 59.8% | 70.6% | 64.8% | +∞%  |

**Massive Improvements**:

1. **Negative Class**: 13.1% → 71.6% F1
   - Now properly detected (73.7% recall vs 8.2%)
   - Class weighting worked!

2. **Positive Class**: 0.0% → 64.8% F1
   - Went from completely missed to well-detected
   - 70.6% recall means most positives found

3. **Neutral Class**: 62.6% → 67.5% F1
   - Maintained good performance
   - Slight recall drop (91.8% → 63.6%) is GOOD
   - No longer overpredicting neutral!

**Balanced Performance** :
- All classes now 60-72% F1
- No class ignored
- No heavy bias to majority class

**Comparison to Other Approaches**:

| Approach | Macro-F1 | Improvement |
|----------|----------|-------------|
| Baseline | 0.2522 | - |
| ICL | 0.2147 | **-14.9%** (worse!) |
| **Fine-tuning** | **0.6794** | **+169%**  |

**Why It Worked**:

1. **Supervised Learning**: Model updated parameters
2. **Class Weighting**: Forced balanced learning
3. **Sufficient Training**: 3 epochs enough to converge
4. **Architecture Match**: Encoder model + fine-tuning = perfect fit

**Remaining Issues**:
- Positive class still lowest F1 (64.8%)
- Could improve with more data or data augmentation
- But overall performance is excellent!

---

# Final Comparison & Conclusions

### Performance Summary Table

| Metric | Baseline | ICL | Fine-tuning | Winner |
|--------|----------|-----|-------------|---------|
| **Macro-F1** | 0.2522 | 0.2147 | **0.6794** |  Fine-tuning |
| **Accuracy** | 0.4630 | 0.4750 | **0.6830** |  Fine-tuning |
| **Negative F1** | 0.1308 | 0.0000 | **0.7155** |  Fine-tuning |
| **Neutral F1** | 0.6260 | 0.6441 | **0.6749** |  Fine-tuning |
| **Positive F1** | 0.0000 | 0.0000 | **0.6478** |  Fine-tuning |
| **Training Time** | 0s | 0s | 167s | - |
| **Inference Time** | ~10s | ~60s | ~10s |  Baseline/FT |



---

## Key Findings

### 1. Fine-Tuning is Essential for Encoder Models
- **169% improvement** over zero-shot baseline
- Encoder models (BERT, DistilBERT) need parameter updates to adapt
- Pretrained knowledge provides good starting point, but task-specific training crucial

### 2. In-Context Learning Fails for DistilBERT
- **Worse than baseline** (-14.9%)
- Encoder-only architecture incompatible with prompt-based learning
- ICL requires decoder models (GPT) or encoder-decoder (T5)
- **Key insight**: Not all modern techniques work with all architectures

### 3. Class Weighting is Critical for Imbalanced Data
- Standard loss → bias toward majority class (baseline behavior)
- Weighted loss → balanced performance across all classes
- **Simple technique, massive impact**: Minority class F1 improved by 400%+

### 4. Evaluation Metrics Matter
- Accuracy alone is misleading (46% looks "okay")
- Macro-F1 reveals true performance (25% shows failure)
- **Always use class-balanced metrics** for imbalanced problems

---

## Practical Recommendations

### When to Use Each Approach

| Approach | Use When | Don't Use When |
|----------|----------|----------------|
| **Zero-shot** | - Quick testing<br>- No labeled data<br>- Exploratory analysis | - Need good performance<br>- Have labeled data<br>- Imbalanced classes |
| **ICL** | - Using GPT/Claude<br>- Very limited data<br>- Need quick deployment | - Using BERT/DistilBERT<br>- Have training data<br>- Need best performance |
| **Fine-tuning** | - Have labeled data<br>- Using encoder models<br>- Need best performance<br>- Can afford training time | - Zero labeled data<br>- Extremely limited compute<br>- Need instant deployment |

### For Imbalanced Classification

**Always**:
- Use class-weighted loss or oversampling
- Evaluate with Macro-F1, not just accuracy
- Monitor per-class metrics
- Check confusion matrix

**Never**:
- Rely solely on accuracy
- Ignore minority class performance
- Use standard loss without considering imbalance
- Skip validation during training

---

## Lessons Learned

1. **Architecture matters**: Different models need different training strategies
2. **Problem understanding matters**: Identified imbalance as key challenge
3. **Simple solutions work**: Class weighting > complex architectures
4. **Negative results are valuable**: ICL failure teaches us something important
5. **Metrics guide decisions**: Right metric reveals true performance

---
